File: src/n64/Animation.cpp
#include <sagas/n64/Animation.hpp>

#include <algorithm>

namespace sagas::n64 {

std::vector<std::optional<Address>> AnimationDecoder::table(Address address, std::size_t count) {
    std::vector<std::optional<Address>> scripts;
    scripts.reserve(count);
    for (std::size_t i = 0; i < count; ++i) {
        const Address word{address.file, address.offset + static_cast<std::uint32_t>(i * 4)};
        scripts.push_back(archive_.u32(word) == 0 ? std::nullopt : archive_.resolve(word));
    }
    return scripts;
}

JointPose AnimationDecoder::pose(const Node& node) {
    JointPose result;
    std::copy(node.rotate.begin(), node.rotate.end(), result.tracks.begin());
    std::copy(node.translate.begin(), node.translate.end(), result.tracks.begin() + 4);
    std::copy(node.scale.begin(), node.scale.end(), result.tracks.begin() + 7);
    return result;
}

void AnimationDecoder::apply(Node& node, const JointPose& pose) {
    std::copy_n(pose.tracks.begin(), 3, node.rotate.begin());
    std::copy_n(pose.tracks.begin() + 4, 3, node.translate.begin());
    std::copy_n(pose.tracks.begin() + 7, 3, node.scale.begin());
}

JointPose AnimationDecoder::sample(Address script, float frame, JointPose initial) {
    enum class Kind { None, Step, Linear, Cubic };
    struct Track {
        Kind kind{Kind::None};
        float base{}, target{}, rate_base{}, rate_target{}, start{}, duration{1};
        bool active{};
        [[nodiscard]] float value(float time) const {
            if (!active) return 0;
            const float length = std::clamp(time - start, 0.0f, duration);
            if (kind == Kind::Step) return length >= duration ? target : base;
            if (kind == Kind::Linear) return base + length * rate_base;
            if (kind != Kind::Cubic || duration == 0) return target;
            const float inv = 1.0f / duration;
            const float x2 = length * length;
            const float inv2 = inv * inv;
            const float x3_inv2 = x2 * length * inv2;
            const float twice = 2.0f * x3_inv2 * inv;
            const float thrice = 3.0f * x2 * inv2;
            const float x2_inv = x2 * inv;
            const float tangent = x3_inv2 - x2_inv;
            return base * ((twice - thrice) + 1.0f) + target * (thrice - twice) +
                   rate_base * ((tangent - x2_inv) + length) + rate_target * tangent;
        }
    };
    std::array<Track, 10> tracks{};
    for (std::size_t i = 0; i < tracks.size(); ++i)
        tracks[i].base = tracks[i].target = initial.tracks[i];
    Address command = script;
    float cursor{};
    std::size_t budget = 16'384;
    while (budget-- && cursor <= frame) {
        const auto word = archive_.u32(command);
        const unsigned opcode = word >> 25;
        const unsigned flags = (word >> 15) & 0x3ffU;
        const float duration = static_cast<float>(word & 0x7fffU);
        command.offset += 4;
        if (opcode == 0) break;
        if (opcode == 1 || opcode == 14) {
            const auto target = archive_.resolve(command);
            if (!target) break;
            command = *target;
            continue;
        }
        if (opcode == 2) { cursor += duration; continue; }
        if (opcode == 13) { command.offset += 4; continue; }
        if (opcode == 15 || opcode == 16) { cursor += duration; continue; }
        if (opcode == 17) {
            for (unsigned bit = 0; bit < 10; ++bit) if (flags & (1U << bit)) command.offset += 4;
            cursor += duration;
            continue;
        }
        const bool values = opcode == 3 || opcode == 4 || opcode == 5 || opcode == 6 ||
                            opcode == 8 || opcode == 9 || opcode == 10 || opcode == 11;
        if (opcode == 7) {
            for (unsigned bit = 0; bit < 10; ++bit) if (flags & (1U << bit)) {
                tracks[bit].rate_target = archive_.f32(command);
                command.offset += 4;
            }
            continue;
        }
        if (!values) break;
        for (unsigned bit = 0; bit < 10; ++bit) if (flags & (1U << bit)) {
            auto& track = tracks[bit];
            track.base = track.target;
            track.target = archive_.f32(command);
            command.offset += 4;
            track.start = cursor;
            track.duration = duration;
            track.active = true;
            if (opcode == 3 || opcode == 4) {
                track.kind = Kind::Linear;
                if (duration != 0) track.rate_base = (track.target - track.base) / duration;
                track.rate_target = 0;
            } else if (opcode == 5 || opcode == 6) {
                track.rate_base = track.rate_target;
                track.rate_target = archive_.f32(command);
                command.offset += 4;
                track.kind = Kind::Cubic;
            } else if (opcode == 8 || opcode == 9) {
                track.rate_base = track.rate_target;
                track.rate_target = 0;
                track.kind = Kind::Cubic;
            } else {
                track.rate_target = 0;
                track.kind = Kind::Step;
            }
        }
        if (opcode == 3 || opcode == 5 || opcode == 8 || opcode == 10) cursor += duration;
    }
    for (std::size_t i = 0; i < tracks.size(); ++i)
        if (tracks[i].active) initial.tracks[i] = tracks[i].value(frame);
    return initial;
}

JointPose AnimationDecoder::sample16(Address script, float frame, JointPose initial) {
    enum class Kind { None, Step, Linear, Cubic };
    struct Track {
        Kind kind{Kind::None};
        float base{}, target{}, rate_base{}, rate_target{}, start{}, duration{1};
        bool active{};
        [[nodiscard]] float value(float time) const {
            if (!active) return target;
            const float length = std::clamp(time - start, 0.0f, duration);
            if (kind == Kind::Step) return duration == 0 || length >= duration ? target : base;
            if (kind == Kind::Linear) return base + length * rate_base;
            if (kind != Kind::Cubic || duration == 0) return target;
            const float inv = 1.0f / duration;
            const float x2 = length * length;
            const float inv2 = inv * inv;
            const float x3_inv2 = x2 * length * inv2;
            const float twice = 2.0f * x3_inv2 * inv;
            const float thrice = 3.0f * x2 * inv2;
            const float x2_inv = x2 * inv;
            const float tangent = x3_inv2 - x2_inv;
            return base * ((twice - thrice) + 1.0f) + target * (thrice - twice) +
                   rate_base * ((tangent - x2_inv) + length) + rate_target * tangent;
        }
    };
    std::array<Track, 10> tracks{};
    for (std::size_t i = 0; i < tracks.size(); ++i)
        tracks[i].base = tracks[i].target = initial.tracks[i];

    const auto read_u16 = [&](Address at) {
        return static_cast<std::uint16_t>(archive_.s16(at));
    };
    const auto converted = [](std::int16_t raw, unsigned track, bool rate) {
        static constexpr std::array<float, 4> value_scale{
            1.0f / 512.0f, 1.0f / 4.0f, 1.0f / 4096.0f, 1.0f / 16384.0f};
        static constexpr std::array<float, 4> rate_scale{
            1.0f / 512.0f, 1.0f / 32.0f, 1.0f / 8192.0f, 1.0f / 16384.0f};
        unsigned kind{};
        if (track >= 4 && track <= 6) kind = 1;
        else if (track >= 7) kind = 2;
        else if (track == 3) kind = 3;
        return static_cast<float>(raw) * (rate ? rate_scale[kind] : value_scale[kind]);
    };

    Address command = script;
    float cursor{};
    std::size_t budget = 65'536;
    while (budget-- && cursor <= frame) {
        const auto word = read_u16(command);
        const unsigned opcode = word >> 11;
        const unsigned flags = (word >> 1) & 0x3ffU;
        const bool toggle = (word & 1U) != 0;
        command.offset += 2;
        if (opcode == 0) break;

        if (opcode == 13) {
            // The loop offset is relative to its own 16-bit word and is
            // expressed in bytes by the original figatree evaluator.
            const auto relative = archive_.s16(command);
            command.offset = static_cast<std::uint32_t>(static_cast<std::int64_t>(command.offset) + relative);
            continue;
        }
        if (opcode == 12) {
            command.offset += 2; // translation interpolation data offset
            continue;
        }

        float duration{};
        if (toggle) {
            duration = static_cast<float>(read_u16(command));
            command.offset += 2;
        }
        if (opcode == 1) {
            cursor += duration;
            continue;
        }
        if (opcode == 11) continue;
        if (opcode == 14) {
            cursor += duration;
            continue;
        }
        if (opcode == 6) {
            for (unsigned bit = 0; bit < tracks.size(); ++bit) if (flags & (1U << bit)) {
                tracks[bit].rate_target = converted(archive_.s16(command), bit, true);
                command.offset += 2;
            }
            continue;
        }
        if (opcode < 2 || opcode > 10) break;

        for (unsigned bit = 0; bit < tracks.size(); ++bit) if (flags & (1U << bit)) {
            auto& track = tracks[bit];
            track.base = track.target;
            track.target = converted(archive_.s16(command), bit, false);
            command.offset += 2;
            track.start = cursor;
            track.duration = duration;
            track.active = true;
            if (opcode == 2 || opcode == 3) {
                track.kind = Kind::Linear;
                track.rate_base = duration != 0 ? (track.target - track.base) / duration : 0;
                track.rate_target = 0;
            } else if (opcode == 4 || opcode == 5) {
                track.rate_base = track.rate_target;
                track.rate_target = converted(archive_.s16(command), bit, true);
                command.offset += 2;
                track.kind = Kind::Cubic;
            } else if (opcode == 7 || opcode == 8) {
                track.rate_base = track.rate_target;
                track.rate_target = 0;
                track.kind = Kind::Cubic;
            } else {
                track.rate_target = 0;
                track.kind = Kind::Step;
            }
        }
        if (opcode == 2 || opcode == 4 || opcode == 7 || opcode == 9) cursor += duration;
    }
    for (std::size_t i = 0; i < tracks.size(); ++i)
        if (tracks[i].active) initial.tracks[i] = tracks[i].value(frame);
    return initial;
}

} // namespace sagas::n64

File: tools/sequence_viewer/animation.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Optional

from .archive import RelocArchive
from .types import Address, Color, MATERIAL_COLORS, Node


@dataclass
class JointPose:
    tracks: list[float] = field(default_factory=lambda: [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 1.0, 1.0])
    flags: int = 0


@dataclass
class MaterialPose:
    tracks: list[float] = field(default_factory=lambda: [0.0, 0.0, 0.0, 1.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0])
    colors: list[Color] = field(default_factory=lambda: [
        Color(255, 255, 255, 255), Color(0, 0, 0, 255), Color(0, 0, 0, 0),
        Color(255, 255, 255, 255), Color(255, 255, 255, 255),
    ])


def pose_from_node(node: Node) -> JointPose:
    return JointPose(tracks=[
        node.rotate[0], node.rotate[1], node.rotate[2], 0.0,
        node.translate[0], node.translate[1], node.translate[2],
        node.scale[0], node.scale[1], node.scale[2],
    ])


def apply_pose(node: Node, pose: JointPose) -> Node:
    return Node(
        depth=node.depth,
        flags=node.flags,
        translate=(pose.tracks[4], pose.tracks[5], pose.tracks[6]),
        rotate=(pose.tracks[0], pose.tracks[1], pose.tracks[2]),
        scale=(pose.tracks[7], pose.tracks[8], pose.tracks[9]),
        display_list=node.display_list,
    )


def animation_table(archive: RelocArchive, address: Address, count: int) -> list[Optional[Address]]:
    scripts: list[Optional[Address]] = []
    for i in range(count):
        word = address.shifted(i * 4)
        scripts.append(None if archive.u32(word) == 0 else archive.resolve(word))
    return scripts


class _FloatTrack:
    __slots__ = ("kind", "base", "target", "rate_base", "rate_target", "start", "duration", "active")

    def __init__(self, value: float = 0.0) -> None:
        self.kind = "none"
        self.base = value
        self.target = value
        self.rate_base = 0.0
        self.rate_target = 0.0
        self.start = 0.0
        self.duration = 1.0
        self.active = False

    def value(self, time: float) -> float:
        if not self.active:
            return self.target
        length = min(max(time - self.start, 0.0), self.duration)
        if self.kind == "step":
            return self.target if self.duration == 0 or length >= self.duration else self.base
        if self.kind == "linear":
            return self.base + length * self.rate_base
        if self.kind != "cubic" or self.duration == 0:
            return self.target
        inv = 1.0 / self.duration
        x2 = length * length
        inv2 = inv * inv
        x3_inv2 = x2 * length * inv2
        twice = 2.0 * x3_inv2 * inv
        thrice = 3.0 * x2 * inv2
        x2_inv = x2 * inv
        tangent = x3_inv2 - x2_inv
        return (self.base * ((twice - thrice) + 1.0) + self.target * (thrice - twice)
                + self.rate_base * ((tangent - x2_inv) + length) + self.rate_target * tangent)


def _blend_color(a: Color, b: Color, amount: float) -> Color:
    def channel(src: int, dst: int) -> int:
        return int(round(src + (dst - src) * amount))
    return Color(channel(a.r, b.r), channel(a.g, b.g), channel(a.b, b.b), channel(a.a, b.a))


class _ColorTrack:
    __slots__ = ("kind", "base", "target", "start", "duration", "active")

    def __init__(self, value: Color) -> None:
        self.kind = "none"
        self.base = value
        self.target = value
        self.start = 0.0
        self.duration = 1.0
        self.active = False

    def value(self, time: float) -> Color:
        if self.kind == "step":
            return self.target if self.duration == 0 or time - self.start >= self.duration else self.base
        amount = 1.0 if self.duration == 0 else min(max((time - self.start) / self.duration, 0.0), 1.0)
        return _blend_color(self.base, self.target, amount)


def _popcount(value: int) -> int:
    return bin(value).count("1")


def sample32(archive: RelocArchive, script: Address, frame: float, initial: JointPose) -> JointPose:
    """Evaluate an AObjEvent32 joint/camera script (ssb-decomp-re src/sys/objanim.c)."""
    tracks = [_FloatTrack(initial.tracks[i]) for i in range(10)]
    command = script
    cursor = 0.0
    budget = 16384
    pose = JointPose(list(initial.tracks), initial.flags)
    while budget > 0 and cursor <= frame:
        budget -= 1
        word = archive.u32(command)
        opcode = word >> 25
        flags = (word >> 15) & 0x3FF
        duration = float(word & 0x7FFF)
        command = command.shifted(4)
        if opcode == 0:
            break
        if opcode in (1, 14):
            target = archive.resolve(command)
            if target is None:
                break
            command = target
            continue
        if opcode == 2:
            cursor += duration
            continue
        if opcode == 13:
            command = command.shifted(4)
            continue
        if opcode == 15:
            pose.flags = flags
            cursor += duration
            continue
        if opcode == 16:
            cursor += duration
            continue
        if opcode == 17:
            command = command.shifted(_popcount(flags) * 4)
            cursor += duration
            continue
        if opcode == 7:
            for bit in range(10):
                if flags & (1 << bit):
                    tracks[bit].rate_target = archive.f32(command)
                    command = command.shifted(4)
            continue
        values = opcode in (3, 4, 5, 6, 8, 9, 10, 11)
        if not values:
            break
        for bit in range(10):
            if not (flags & (1 << bit)):
                continue
            track = tracks[bit]
            track.base = track.target
            track.target = archive.f32(command)
            command = command.shifted(4)
            track.start = cursor
            track.duration = duration
            track.active = True
            if opcode in (3, 4):
                track.kind = "linear"
                track.rate_base = (track.target - track.base) / duration if duration else 0.0
                track.rate_target = 0.0
            elif opcode in (5, 6):
                track.rate_base = track.rate_target
                track.rate_target = archive.f32(command)
                command = command.shifted(4)
                track.kind = "cubic"
            elif opcode in (8, 9):
                track.rate_base = track.rate_target
                track.rate_target = 0.0
                track.kind = "cubic"
            else:
                track.rate_target = 0.0
                track.kind = "step"
        if opcode in (3, 5, 8, 10):
            cursor += duration
    for i, track in enumerate(tracks):
        if track.active:
            pose.tracks[i] = track.value(frame)
    return pose


def sample_material(archive: RelocArchive, script: Address, frame: float,
                    initial: MaterialPose) -> MaterialPose:
    tracks = [_FloatTrack(initial.tracks[i]) for i in range(10)]
    colors = [_ColorTrack(initial.colors[i]) for i in range(len(MATERIAL_COLORS))]
    command = script
    cursor = 0.0
    budget = 16384
    pose = MaterialPose(list(initial.tracks), list(initial.colors))

    def packed_color(at: Address) -> Color:
        packed = archive.u32(at)
        return Color((packed >> 24) & 255, (packed >> 16) & 255, (packed >> 8) & 255, packed & 255)

    while budget > 0 and cursor <= frame:
        budget -= 1
        word = archive.u32(command)
        opcode = word >> 25
        flags = (word >> 15) & 0x3FF
        duration = float(word & 0x7FFF)
        command = command.shifted(4)
        if opcode == 0:
            break
        if opcode in (1, 14):
            target = archive.resolve(command)
            if target is None:
                break
            command = target
            continue
        if opcode == 2:
            cursor += duration
            continue
        if opcode == 12:
            continue
        if opcode == 13:
            command = command.shifted(4)
            continue
        if opcode in (15, 16):
            cursor += duration
            continue
        if opcode == 17:
            command = command.shifted(_popcount(flags) * 4)
            cursor += duration
            continue
        if 18 <= opcode <= 21:
            for bit in range(len(colors)):
                if flags & (1 << bit):
                    track = colors[bit]
                    track.base = track.target
                    track.target = packed_color(command)
                    command = command.shifted(4)
                    track.start = cursor
                    track.duration = duration
                    track.kind = "step" if opcode in (18, 19) else "linear"
                    track.active = True
            if opcode in (18, 20):
                cursor += duration
            continue
        if opcode == 22:
            command = command.shifted(_popcount(flags & 0x1F) * 4)
            cursor += duration
            continue
        if opcode == 7:
            for bit in range(10):
                if flags & (1 << bit):
                    tracks[bit].rate_target = archive.f32(command)
                    command = command.shifted(4)
            continue
        values = opcode in (3, 4, 5, 6, 8, 9, 10, 11)
        if not values:
            break
        for bit in range(10):
            if not (flags & (1 << bit)):
                continue
            track = tracks[bit]
            track.base = track.target
            track.target = archive.f32(command)
            command = command.shifted(4)
            track.start = cursor
            track.duration = duration
            track.active = True
            if opcode in (3, 4):
                track.kind = "linear"
                track.rate_base = (track.target - track.base) / duration if duration else 0.0
                track.rate_target = 0.0
            elif opcode in (5, 6):
                track.rate_base = track.rate_target
                track.rate_target = archive.f32(command)
                command = command.shifted(4)
                track.kind = "cubic"
            elif opcode in (8, 9):
                track.rate_base = track.rate_target
                track.rate_target = 0.0
                track.kind = "cubic"
            else:
                track.rate_target = 0.0
                track.kind = "step"
        if opcode in (3, 5, 8, 10):
            cursor += duration
    for i, track in enumerate(tracks):
        if track.active:
            pose.tracks[i] = track.value(frame)
    for i, track in enumerate(colors):
        if track.active:
            pose.colors[i] = track.value(frame)
    return pose


def _figatree_scale(raw: int, track: int, rate: bool) -> float:
    value_scale = (1.0 / 512.0, 1.0 / 4.0, 1.0 / 4096.0, 1.0 / 16384.0)
    rate_scale = (1.0 / 512.0, 1.0 / 32.0, 1.0 / 8192.0, 1.0 / 16384.0)
    kind = 0
    if 4 <= track <= 6:
        kind = 1
    elif track >= 7:
        kind = 2
    elif track == 3:
        kind = 3
    table = rate_scale if rate else value_scale
    return float(raw) * table[kind]


def sample16(archive: RelocArchive, script: Address, frame: float,
             initial: Optional[JointPose] = None) -> JointPose:
    """Evaluate a figatree AObjEvent16 fighter motion script."""
    if initial is None:
        initial = JointPose()
    tracks = [_FloatTrack(initial.tracks[i]) for i in range(10)]
    command = script
    cursor = 0.0
    budget = 65536
    pose = JointPose(list(initial.tracks), initial.flags)
    while budget > 0 and cursor <= frame:
        budget -= 1
        word = archive.u16(command)
        opcode = word >> 11
        flags = (word >> 1) & 0x3FF
        toggle = (word & 1) != 0
        command = command.shifted(2)
        if opcode == 0:
            break
        if opcode == 13:
            relative = archive.s16(command)
            command = Address(command.file, command.offset + relative)
            continue
        if opcode == 12:
            command = command.shifted(2)
            continue
        duration = 0.0
        if toggle:
            duration = float(archive.u16(command))
            command = command.shifted(2)
        if opcode == 1:
            cursor += duration
            continue
        if opcode == 11:
            continue
        if opcode == 14:
            pose.flags = flags
            cursor += duration
            continue
        if opcode == 6:
            for bit in range(10):
                if flags & (1 << bit):
                    tracks[bit].rate_target = _figatree_scale(archive.s16(command), bit, True)
                    command = command.shifted(2)
            continue
        if opcode < 2 or opcode > 10:
            break
        for bit in range(10):
            if not (flags & (1 << bit)):
                continue
            track = tracks[bit]
            track.base = track.target
            track.target = _figatree_scale(archive.s16(command), bit, False)
            command = command.shifted(2)
            track.start = cursor
            track.duration = duration
            track.active = True
            if opcode in (2, 3):
                track.kind = "linear"
                track.rate_base = (track.target - track.base) / duration if duration else 0.0
                track.rate_target = 0.0
            elif opcode in (4, 5):
                track.rate_base = track.rate_target
                track.rate_target = _figatree_scale(archive.s16(command), bit, True)
                command = command.shifted(2)
                track.kind = "cubic"
            elif opcode in (7, 8):
                track.rate_base = track.rate_target
                track.rate_target = 0.0
                track.kind = "cubic"
            else:
                track.rate_target = 0.0
                track.kind = "step"
        if opcode in (2, 4, 7, 9):
            cursor += duration
    for i, track in enumerate(tracks):
        if track.active:
            pose.tracks[i] = track.value(frame)
    return pose
